### PR TITLE
feat: implements block height check in PrepareProposal to avoid unexpected errors at height 1

### LINKED ISSUE #1899: Test for the ability to include transactions in the first block
Currently, we have never tested that transactions can actually be included in the first block. We also have issues with accessing the state in PrepareProposal, that's why we have these lines

https://github.com/celestiaorg/celestia-app/blob/9bf0cf1dd9ce31a3fecb51310c3913820b21a8c2/app/square_size.go#L11-L19

If we can't actually process txs in the first block, and there ends up being transactions in the mempool, then we could end up with a "frustrated" start to the chain where we won't be able to create a block until a block producer creates an empty one.

### LINKED ISSUE #2221: Implementing block height check in PrepareProposal for improved behavior
The following suggestion is related to #1899, and is originally suggested by @evan-forbes in this [comment](https://github.com/celestiaorg/celestia-app/pull/2220#pullrequestreview-1565761861)  

> we don't have to handle it all in this PR, but after we add the hacky if statement protection in PrepareProposal, we could also add a different test that waits for block height 2 and then ensures that the first block is empty

### PR DESCRIPTION
**Pull request by staheri14**
## Overview
Closes #1899 
Closes #2221

## Checklist

- [ ] New and updated code has appropriate documentation
- [ ] New and updated code has new and/or updated testing
- [ ] Required CI checks are passing
- [ ] Visual proof for any user facing features like CLI or documentation updates
- [ ] Linked issues closed with keywords

### CODE REVIEW

File: specs/src/specs/cat_pool.md [hunk 1]
```diff
@@ -3,25 +3,25 @@
 - 01.12.2022 | Initial specification (@cmwaters)
 - 09.12.2022 | Add Push/Pull mechanics (@cmwaters)
 
-### Outline
```
[the remark this answers is not in the document]
Comment (rootulp) [reply specs/src/specs/cat_pool.md:6]: Thanks for fixing! We could update the PR description to say "Closes https://github.com/celestiaorg/celestia-app/issues/2227" or extract the lint fix to a separate PR. Defer to you

File: app/prepare_proposal.go [hunk 2]
```diff
@@ -32,7 +32,23 @@ func (app *App) PrepareProposal(req abci.RequestPrepareProposal) abci.ResponsePr
 		ante.DefaultSigVerificationGasConsumer,
 		app.IBCKeeper,
 	)
-	txs := FilterTxs(sdkCtx, handler, app.txConfig, req.BlockData.Txs)
+
+	var txs [][]byte
+	if app.LastBlockHeight() == 0 {
```
Comment (evan-forbes) [inline app/prepare_proposal.go:37]: [optional]

I'll defer to the author if its needed or not, but it might be nice to briefly explain this or link to an issue that does for more context

**Review by evan-forbes — APPROVED** [review]
dope! thanks, can we also open an issue to remove this after mainnet?

File: specs/src/specs/cat_pool.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (staheri14) [reply specs/src/specs/cat_pool.md]: Thanks for mentioning the GH issue, I cut a separate PR for it, please see https://github.com/celestiaorg/celestia-app/pull/2242

File: app/prepare_proposal.go [hunk 3]
```diff
@@ -32,7 +32,23 @@ func (app *App) PrepareProposal(req abci.RequestPrepareProposal) abci.ResponsePr
...
+		// the first block is expected to be empty; hence return an empty set of txs
+		txs = make([][]byte, 0)
+		if len(req.BlockData.Txs) != 0 {
+			// if the consensus layer sends non-empty set of transactions for
+			// block height 1, log it
+			app.Logger().Info(
+				"non-empty txs received from the consensus layer for block height 1",
+				"invalid",
+				len(req.BlockData.Txs),
```
Comment (SweeXordious) [inline app/prepare_proposal.go:46]:
```suggestion
				"non-empty txs received from the consensus layer for block height 1",
				"numberOfTransactions",
				len(req.BlockData.Txs),
```

**Review by evan-forbes — DISMISSED** [review]

File: app/prepare_proposal.go (hunk 2 above)
Comment (cmwaters) [reply app/prepare_proposal.go:37]: Yeah since this is somewhat of an unusual hack I think it would be good to point out what why this is needed

File: app/prepare_proposal.go [hunk 4]
```diff
@@ -32,7 +32,23 @@ func (app *App) PrepareProposal(req abci.RequestPrepareProposal) abci.ResponsePr
...
-	txs := FilterTxs(sdkCtx, handler, app.txConfig, req.BlockData.Txs)
+
+	var txs [][]byte
+	if app.LastBlockHeight() == 0 {
+		// the first block is expected to be empty; hence return an empty set of txs
+		txs = make([][]byte, 0)
+		if len(req.BlockData.Txs) != 0 {
+			// if the consensus layer sends non-empty set of transactions for
+			// block height 1, log it
+			app.Logger().Info(
```
Comment (cmwaters) [inline app/prepare_proposal.go:43]: Why do we want to log this? i.e. what value does it bring?

**Review by cmwaters — DISMISSED** [review]
LGTM just have one non-blocking question

File: app/prepare_proposal.go (hunk 3 above)
Comment (staheri14) [reply app/prepare_proposal.go]: Thanks, fixed it in https://github.com/celestiaorg/celestia-app/pull/2233/commits/93692c5d4876e18d3c4f366e8d9126df608679c4.

File: app/prepare_proposal.go (hunk 2 above)
Comment (staheri14) [reply app/prepare_proposal.go:49]: Sure, please see the added explanations https://github.com/celestiaorg/celestia-app/pull/2233/commits/93692c5d4876e18d3c4f366e8d9126df608679c4.

**Review by evan-forbes — APPROVED** [review]

**Comment by staheri14** [comment]
> also open an issue to remove this after mainnet

Sure, done https://github.com/celestiaorg/celestia-app/issues/2271

File: app/test/block_production_test.go [hunk 5]
```diff
@@ -55,3 +57,35 @@ func (s *BlockProductionTestSuite) Test_BlockOneTransactionNonInclusion() {
...
+	// wait until height 1 before posting transactions
+	// otherwise tx submission will fail
+	time.Sleep(1 * s.timeoutCommit)
+	// send some transactions, these should be included in the second block
+	_, err := s.cctx.PostData(s.accounts[0], flags.BroadcastBlock, appns.RandomBlobNamespace(), tmrand.Bytes(100000))
+	require.NoError(err)
+
+	// wait for 2*s.timeoutCommit+1*time.Second to ensure that the node is
+	// at height 2
+	_, err = s.cctx.WaitForHeightWithTimeout(2, 2*s.timeoutCommit+1*time.Second)
```
Comment (rootulp) [inline app/test/block_production_test.go:73]: [question] if this only cares about waiting for block height 2, is that specific time necessary? Very minor but It's difficult to read. Put another way, this seems to work:

```suggestion
	// Wait for block height 2.
	_, err = s.cctx.WaitForHeightWithTimeout(2, s.timeoutCommit)
	require.NoError(err)
```

File: app/test/block_production_test.go [hunk 6]
```diff
@@ -55,3 +57,35 @@ func (s *BlockProductionTestSuite) Test_BlockOneTransactionNonInclusion() {
 	require.Error(err) // change this to require.NoError(err) to see the error
 	require.ErrorContains(err, "rpc error: code = Unknown desc = codespace sdk code 18: invalid request: failed to load state at height 0; no commit info found (latest height: 0)")
 }
+
+// Test_FirstBlockIsEmpty tests whether the first block is empty.
+func (s *BlockProductionTestSuite) Test_FirstBlockIsEmpty() {
+	require := s.Require()
+	// wait until height 1 before posting transactions
+	// otherwise tx submission will fail
+	time.Sleep(1 * s.timeoutCommit)
```
Comment (rootulp) [inline app/test/block_production_test.go:66]: [question][not blocking] it's not immediately obvious what the relationship is between timeout commit and block height 1. If this only cares about waiting for block height 1 then it could be

```suggestion
	// Wait until block height 1 before posting transactions otherwise tx
	// submission will fail.
	_, err := s.cctx.WaitForHeightWithTimeout(1, s.timeoutCommit)
	require.NoError(err)
```

**Review by rootulp — APPROVED** [review]

File: app/prepare_proposal.go (hunk 4 above)
Comment (cmwaters) [reply app/prepare_proposal.go:54]: But we can conclude that it could be the case right? I'm just trying to work out what value it provides for node operators

Comment (staheri14) [reply app/prepare_proposal.go:54]: > But we can conclude that it could be the case right?

We expect that it would not be the case at all unless we observe this logged info. 

> I'm just trying to work out what value it provides for node operators

I understand your point. The current log doesn't tell node operators much unless they're interested in the problem of transactions in the mempool at height 0. However, having this log could help us check if this happens during local tests. I don't have a strong opinion on this though. If you think it's better, we can remove the log. Just let me know.

**Comment by staheri14** [comment]
Oops, my bad. The auto-merge was on, so the PR got merged. I'll create a new PR to take care of the comments.

File: app/test/block_production_test.go (hunk 6 above)
Comment (staheri14) [reply app/test/block_production_test.go:66]: Both methods work well: either sleeping for the timeoutCommit or waiting for height 1. However, for a clearer understanding, I've chosen to explicitly wait, and I've incorporated your suggestion in this [PR](https://github.com/celestiaorg/celestia-app/pull/2284). Thanks.

File: app/test/block_production_test.go (hunk 5 above)
Comment (staheri14) [reply app/test/block_production_test.go:73]: Thanks, your comment is addressed in this [PR](https://github.com/celestiaorg/celestia-app/pull/2284).

File: app/prepare_proposal.go (hunk 4 above)
Comment (staheri14) [reply app/prepare_proposal.go:54]: Created this PR to address your comment https://github.com/celestiaorg/celestia-app/pull/2285
